apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-metrics-exporter/src/main/java/co/elastic/apm/agent/otelmetricexport/MetricSetGenerator.java

```diff
@@ -0,0 +1,288 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.otelmetricexport;
+
+import co.elastic.apm.agent.report.Reporter;
+import co.elastic.apm.agent.report.serialize.DslJsonSerializer;
+import co.elastic.apm.agent.sdk.logging.Logger;
+import co.elastic.apm.agent.sdk.logging.LoggerFactory;
+import com.dslplatform.json.BoolConverter;
+import com.dslplatform.json.DslJson;
+import com.dslplatform.json.JsonWriter;
+import com.dslplatform.json.Nullable;
+import com.dslplatform.json.NumberConverter;
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.AttributeType;
+import io.opentelemetry.api.common.Attributes;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.dslplatform.json.JsonWriter.ARRAY_END;
+import static com.dslplatform.json.JsonWriter.ARRAY_START;
+import static com.dslplatform.json.JsonWriter.COMMA;
+import static com.dslplatform.json.JsonWriter.OBJECT_END;
+import static com.dslplatform.json.JsonWriter.OBJECT_START;
+
+class MetricSetGenerator {
+
+    private static final Logger logger = LoggerFactory.getLogger(MetricSetGenerator.class);
+
+    private static final byte NEW_LINE = '\n';
+
+    private static final int INITIAL_BUFFER_SIZE = 2048;
+
+    private static final DslJson<Object> DSL_JSON = new DslJson<>(new DslJson.Settings<>());
+
+    private final StringBuilder replaceBuilder;
+    private final JsonWriter jw;
+    private boolean anySamplesWritten;
+
+    public MetricSetGenerator(Attributes attributes, CharSequence instrumentationScopeName, long epochMicros, StringBuilder replaceBuilder) {
```

Comment:
[For Reviewer]:
We now have metricset serialization in three places: The agent metrics, Micrometer and Otel. The `MetricSetGenerator` class is designed to allow a refactoring in the future to have the serialization logic in one place again.
The idea is to extract a common superclass which delegates just the `tags` generation to the subclasses.

If you agree that this is a good Idea, I'll create a tech-debt issue for this after this PR is merged.

Reply:
I think that's a very good idea, the duplication with micrometer metrics was somehow acceptable, but now it's clear we really need to prioritize this. As part of this refactor, we should include properly testing this class independently from the otel/micrometer/internal implementations as it is currently only tested indirectly.